### PR TITLE
PMM-14763 RDS functionality to refresh new name.

### PR DESCRIPTION
**Pull request by JiriCtvrtka**
PMM-14763

FB: https://github.com/Percona-Lab/pmm-submodules/pull/4373

### CODE REVIEW

**Comment by JiriCtvrtka** [comment]
@copilot review

**Review by maxkondr — DISMISSED** [review]

File: enhanced/scraper.go [hunk 1]
```diff
@@ -65,6 +71,10 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
 	allMetrics := make(map[string]map[time.Time][]prometheus.Metric) // ResourceID -> event timestamp -> metrics
 	allMessages := make(map[string]map[time.Time]string)             // ResourceID -> event timestamp -> message
 
+	if err := s.refreshResourceIDs(ctx); err != nil {
```
Comment (ademidoff) [inline enhanced/scraper.go:78]: This puts an RDS API call (N of them, see below) on the scrape hot path that fires every 2–60 s. `start()` wraps the scrape in `context.WithTimeout(ctx, interval)`, so if `DescribeDBInstances` is slow or throttled, the scrape window shrinks and the actual `FilterLogEvents` work can be starved. Resource IDs change very rarely (blue/green swap, restore-to-point-in-time) — checking every 2 s is wildly over-provisioned. Suggest moving refresh to a separate goroutine with a 5–15 min ticker.

File: enhanced/scraper.go [hunk 2]
```diff
@@ -162,6 +172,31 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
 	return resMetrics, resMessages
 }
 
+func (s *scraper) refreshResourceIDs(ctx context.Context) error {
+	for instanceIndex, instance := range s.instances {
+		resourceID, err := s.resourceIDResolver.ResourceID(ctx, instance.Instance)
```
Comment (ademidoff) [inline enhanced/scraper.go]: N+1 problem: one `DescribeDBInstances` call per instance per scrape. The existing code in `sessions.New()` (sessions/sessions.go around lines 117–146) already proves the bulk pattern — paginated `DescribeDBInstances` with no `DBInstanceIdentifier` filter returns up to 100 instances per page. Reusing that pattern makes this O(1) calls per session instead of O(N), and avoids burning the shared per-region RDS API throttle budget.

File: enhanced/scraper.go [hunk 3]
```diff
@@ -162,6 +172,31 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
...
+		if err != nil {
+			return fmt.Errorf("failed to refresh resource ID for %s: %w", instance.Instance, err)
```
Comment (ademidoff) [inline enhanced/scraper.go]: Early return on the first error skips every remaining instance in this cycle. A single transient throttle on `instances[0]` blocks rename detection for everything else. Prefer logging per-instance and continuing the loop.

File: enhanced/scraper.go [hunk 4]
```diff
@@ -162,6 +172,31 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
...
+		}
+
+		if resourceID == "" || resourceID == instance.ResourceID {
```
Comment (ademidoff) [inline enhanced/scraper.go]: `resourceID == ""` is silently treated as "unchanged", but it actually means "instance not present in the DescribeDBInstances response" (deleted, wrong identifier, permissions issue). Worth at least a `level.Warn` so operators can distinguish a deleted instance from an unchanged one — otherwise the exporter will keep polling a dead log stream forever with no signal.

File: enhanced/scraper.go [hunk 5]
```diff
@@ -162,6 +172,31 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
...
+		}
+
+		level.Info(s.logger).Log(
+			"msg", "RDS resource ID changed.",
+			"region", instance.Region,
+			"instance", instance.Instance,
+			"old_resource_id", instance.ResourceID,
+			"new_resource_id", resourceID,
+		)
+		s.instances[instanceIndex].ResourceID = resourceID
```
Comment (ademidoff) [inline enhanced/scraper.go]: In-place mutation of `s.instances` and `s.logStreamNames`. Safe today because both the first scrape (called from `NewCollector`) and the ticker loop run serially in a single goroutine, but the invariant isn't documented. If someone later parallelises scrapes or exposes `s.instances`, this races silently. A short comment marking the single-goroutine assumption — or a small mutator helper — would help.

File: sessions/sessions.go [hunk 6]
```diff
@@ -45,6 +45,37 @@ type Sessions struct {
...
+		svc: rds.NewFromConfig(cfg),
+	}
+}
+
+// ResourceID returns the current RDS resource ID for a DB instance identifier.
+func (r *ResourceIDResolver) ResourceID(ctx context.Context, instanceID string) (string, error) {
+	output, err := r.svc.DescribeDBInstances(ctx, &rds.DescribeDBInstancesInput{
+		DBInstanceIdentifier: aws.String(instanceID),
+		Filters:              nil,
+		Marker:               nil,
```
Comment (ademidoff) [inline sessions/sessions.go]: `Filters: nil, Marker: nil, MaxRecords: nil` is noise — these are already the zero values. Compare with the existing call at line 122 of this file which only sets the fields it actually uses.

File: enhanced/scraper_test.go [hunk 7]
```diff
@@ -92,6 +92,61 @@ func TestScraper(t *testing.T) {
...
+		testDisallowUnknownFields: false,
+	}
+
+	err := scraper.refreshResourceIDs(t.Context())
+
+	require.NoError(t, err)
+	assert.Equal(t, "new-resource-id", scraper.instances[0].ResourceID)
+	assert.Equal(t, "new-resource-id", scraper.logStreamNames[0])
+	assert.Equal(t, "same-resource-id", scraper.instances[1].ResourceID)
+	assert.Equal(t, "same-resource-id", scraper.logStreamNames[1])
```
Comment (ademidoff) [inline enhanced/scraper_test.go]: Only the happy path is covered. Missing cases:
- Resolver returns an error → exercises the early-return-aborts-loop behaviour (which itself is the bug flagged in scraper.go).
- `resourceID == ""` → instance not found in RDS response.
- All-unchanged no-op path.

Also: adding a call counter to `fakeResourceIDResolver` would let you assert exactly N calls — which directly surfaces the N+1 issue.

**Review by ademidoff — CHANGES_REQUESTED** [review]
The goal here is right — without resource-ID refresh, enhanced metrics break silently after a blue/green cutover. But the cost model is a regression: this adds O(N) `DescribeDBInstances` calls on the scrape hot path every 2–60 s. With ~30 instances on a 1 s monitoring interval, that's ~900 RDS API calls/min per session just to detect a rename — likely to throttle and starve the actual `FilterLogEvents` call. Suggest decoupling the refresh to a separate slow ticker (5–15 min) that reuses the bulk-paginated `DescribeDBInstances` pattern already in `sessions.New()`. Detailed comments inline.

File: enhanced/scraper.go (hunk 2 above)
Comment (ademidoff) [reply enhanced/scraper.go]: API calls cost money, I think we have to be careful here.

**Comment by JiriCtvrtka** [comment]
Updated: resource ID refresh now runs at most once every 5 minutes and uses bulk paginated DescribeDBInstances instead of per-instance calls on every scrape. 

30 instances: typically 1 call / 5 min
100 instances: typically 1 call / 5 min
101+ instances: typically 2+ calls / 5 min, depending on pagination

I also added edge-case tests for resolver errors, missing resource IDs, unchanged IDs, refresh throttling, and fixed the lint issues.

**Comment by JiriCtvrtka** [comment]
@copilot review

File: enhanced/scraper.go [hunk 8]
```diff
@@ -162,6 +176,55 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
 	return resMetrics, resMessages
 }
 
+func (s *scraper) refreshResourceIDsIfNeeded(ctx context.Context) error {
```
Comment (ademidoff) [inline enhanced/scraper.go]:
```suggestion
func (s *scraper) refreshResourceIDs(ctx context.Context) error {
```
We always try to optimize performance, IMO it's redundant to mention optimization in method names.

File: sessions/sessions.go [hunk 9]
```diff
@@ -45,6 +45,51 @@ type Sessions struct {
...
+
+	var marker *string
+
+	for {
+		output, err := r.svc.DescribeDBInstances(ctx, &rds.DescribeDBInstancesInput{
+			DBInstanceIdentifier: nil,
+			Filters:              nil,
+			Marker:               marker,
+			MaxRecords:           nil,
+		})
```
Comment (ademidoff) [inline sessions/sessions.go]:
```suggestion
		output, err := r.svc.DescribeDBInstances(ctx, &rds.DescribeDBInstancesInput{
			Marker:               marker,
		})
```

nit: not super important, but why don't we make it simple?

File: enhanced/scraper.go [hunk 10]
```diff
@@ -14,13 +14,21 @@ import (
 	"github.com/percona/rds_exporter/sessions"
 )
 
+type resourceIDResolver interface {
+	ResourceIDs(ctx context.Context) (map[string]string, error)
+}
+
+const resourceIDRefreshInterval = 5 * time.Minute
```
Comment (ademidoff) [inline enhanced/scraper.go:21]: Let's discuss this interval on the next occasion.

File: sessions/sessions.go (hunk 9 above)
Comment (JiriCtvrtka) [reply sessions/sessions.go]: Applied and added nolint rule.

File: enhanced/scraper.go [hunk 11]
```diff
@@ -162,6 +176,55 @@ func (s *scraper) scrape(ctx context.Context) (map[string][]prometheus.Metric, m
 	return resMetrics, resMessages
 }
 
+func (s *scraper) refreshResourceIDs(ctx context.Context) error {
+	if time.Now().Before(s.nextResourceIDRefresh) {
```
Comment (maxkondr) [inline enhanced/scraper.go:180]: why not `time.Ticker`?

Comment (JiriCtvrtka) [reply enhanced/scraper.go:180]: Lazy refresh can save RDS API calls when /enhanced is not being scraped. It also do not need extra synchronization for scraper state.

**Review by ademidoff — APPROVED** [review]

**Review by maxkondr — APPROVED** [review]
